### ISSUE TITLE
Composite Caption LineList Creation ?

### ISSUE THREAD

**Issue by nmonet** [Wangt-CN/DisCo#38]
Hi ! Thanks for your Disco paper and explanation for the TSV file preparation.

In the composite yaml file, you have a 'caption linelist' file which is used.
caption_linelist: train_TiktokDance-coco-single_person-Lindsey_0411_youtube-SHHQ-1.0-deepfashion2-laion_human-masks-single_cap.caption.linelist.tsv
Could you explain how you make this file ?

**Comment by uk9921** [comment]
I looked into `xxx.caption.linelist.tsv` file, which only has two columns. 
```python
0	0
1	0
2	0
3	0
4	0
5	0
6	0
7	0
8	0
9	0
10	0
11	0
12	0
...
``` 
Please refer to the code here.
https://github.com/Wangt-CN/DisCo/blob/main/dataset/tsv_dataset.py#L490
```python
if self.is_composite:
    assert op.isfile(self.cap_linelist_file)
    self.cap_line_list = [
        int(row[2]) for row in tsv_reader(self.cap_linelist_file)]
    self.img_line_list = [i for i in range(len(self.cap_line_list))]
elif self.cap_linelist_file:
    line_list = load_box_linelist_file(self.cap_linelist_file)
    self.img_line_list = line_list[0]
    self.cap_line_list = line_list[1]
else:
    # one caption per image/video
    self.img_line_list = [i for i in range(self.cap_tsv.num_rows())]
    self.cap_line_list = [0 for i in range(self.cap_tsv.num_rows())]
``` 

On the left side, there is a column called `img_line_list` which seems to range from 0 to the number of data rows. On the right side, there is a column called `cap_line_list` that remains as 0.

Since there is no official documentation provided regarding `xxx.caption.linelist`, it seems that we only need to make a simple modification here:

```python
if self.is_composite:  # false
    assert op.isfile(self.cap_linelist_file)
    self.cap_line_list = [
        int(row[2]) for row in tsv_reader(self.cap_linelist_file)]
    self.img_line_list = [i for i in range(len(self.cap_line_list))]
elif self.cap_linelist_file:  # official data training in
    line_list = load_box_linelist_file(self.cap_linelist_file)
    self.img_line_list = line_list[0]
    self.cap_line_list = line_list[1]
elif self.cap_tsv is not None:
    # one caption per image/video
    self.img_line_list = [i for i in range(self.cap_tsv.num_rows())]
    self.cap_line_list = [0 for i in range(self.cap_tsv.num_rows())]
else:  # w/o cap_line and cap_tsv, user data training in,
    self.img_line_list = [i for i in range(self.visual_tsv.num_rows())]
    self.cap_line_list = [0 for i in range(self.visual_tsv.num_rows())]
``` 
I will give it a try first, and if there is any progress, I will update it here.

**Comment by yoimisan** [comment]
I make a little modification base on the code to make the code goes ok.

First in `./tsv_example/create_custom_dataset_tsvs.py` file. I add some code to generate caption and caption line idx. The caption is just "", and the line idx is just as the @uk9921 says above.

```python
    # process captions & caption line list
    def gen_row_captions(image_path_list):
        for i, image_path in enumerate(image_path_list):
            image_name = image_path.split("/")[-1]
            video_key = image_path.split("/")[-2]
            image_key = f"Custom_{video_key}_{image_name}"
            row = [image_key, '[{"captions": ""}]']
            yield(row)
    tsv_writer(gen_row_captions(image_path_list), f"{args.output_folder}/{split}_captions.tsv")

    def gen_row_captions_line(image_path_list): 
        for i, image_path in enumerate(image_path_list):
            row = [i, 0]
            yield(row)
    tsv_writer(gen_row_captions_line(image_path_list), f"{args.output_folder}/{split}_TiktokDance-poses-masks.caption.linelist.tsv")
```

After that, in `./dataset/tsv_cond_dataset.py`, in function `TsvCondImgCompositeDataset::get_reference_frame_idx(self, img_idx)` 

```python
# Comment out this line 
# start_end = [item + img_idx for item in start_end]
``` 

This is because the `vid2line` file generated by `create_custom_dataset_tsvs.py` records the start offset and end offset of one video in image list. So adding img_idx will cause out of range error. 

---

The bash command line is as follow. At least, it works ok on this command line. :)

```bash
WANDB_ENABLE=0 AZFUSE_USE_FUSE=0 NCCL_ASYNC_ERROR_HANDLING=0 python finetune_sdm_yaml.py --cf config/ref_attn_clip_combine_controlnet/tiktok_S256L16_xformers_tsv.py \
--do_train --root_dir /root/DisCo-main/ \
--local_train_batch_size 48 \
--local_eval_batch_size 48 \
--log_dir exp/tiktok_ft \
--epochs 20 --deepspeed \
--eval_step 500 --save_step 500 \
--gradient_accumulate_steps 1 \
--learning_rate 2e-4 --fix_dist_seed --loss_target "noise" \
--train_yaml /root/autodl-tmp/processed_dataset/train.yaml \
--val_yaml /root/autodl-tmp/processed_dataset/new10val.yaml \
--unet_unfreeze_type "all" \
--guidance_scale 3 \
--refer_sdvae \
--ref_null_caption False \
--combine_clip_local --combine_use_mask \
--conds "poses" "masks" \
--stage1_pretrain_path /root/autodl-tmp/mp_rank_00_model_states.pt 
--drop_ref 0.05 
--guidance_scale 1.5 
```
